src/XMakeBuildEngine/Definition/Project.cs

```diff
@@ -1033,6 +1034,232 @@ public static string GetEvaluatedItemIncludeEscaped(ProjectItemDefinition item)
+        /// </returns>
+        public List<ProvenanceResult> GetItemProvenance(string itemToMatch)
+        {
+            return GetItemProvenance(itemToMatch, _data.EvaluatedItemElements);
+        }
+
+        /// <summary>
+        /// Overload of <see cref="GetItemProvenance(string)"/>
+        /// </summary>
+        /// <param name="itemToMatch">The string to perform matching against</param>
+        /// <param name="itemType">The item type to constrain the search in</param>
+        /// <returns></returns>
+        public List<ProvenanceResult> GetItemProvenance(string itemToMatch, string itemType)
+        {
+            return GetItemProvenance(itemToMatch, _data.EvaluatedItemElements.Where(i => i.ItemType.Equals(itemType)));
+        }
+
+        /// <summary>
+        /// Overload of <see cref="GetItemProvenance(string)"/>
+        /// </summary>
+        /// <param name="item"> 
+        /// The ProjectItem object that indicates: the itemspec to match and the item type to constrain the search in.
+        /// The search is also constrained on item elements appearing before the item element that produced this <paramref name="item"/>.
+        /// The element that produced this <paramref name="item"/> is included in the results.
+        /// </param>
+        /// <returns></returns>
+        public List<ProvenanceResult> GetItemProvenance(ProjectItem item)
+        {
+            var itemElementsAbove = _data.EvaluatedItemElements
+                                            .Where(i => i.ItemType.Equals(item.ItemType))
+                                            .TakeWhile(i => i != item.Xml)
+                                            .ToList();
+            itemElementsAbove.Add(item.Xml);
+
+            return GetItemProvenance(item.EvaluatedInclude, itemElementsAbove);
+        }
+
+
+        private List<ProvenanceResult> GetItemProvenance(string itemToMatch, IEnumerable<ProjectItemElement> projectItemElements )
+        {
+            return
+                projectItemElements.Select(i => ComputeProvenanceResult(itemToMatch, i))
+                    .Where(r => r != null)
+                    .ToList();
+        }
+
+        private ProvenanceResult ComputeProvenanceResult(string itemToMatch, ProjectItemElement itemElement)
+        {
+            var expander = new Expander<ProjectProperty, ProjectItem>(_data.Properties, _data.Items);
+            Func<IElementLocation, Func<string, ExpanderOptions, string>> expandForXmlLocation = (l) => (s, o) => expander.ExpandIntoStringLeaveEscaped(s, o, l);
+
+            var includeResult = ComputeProvenanceResult(itemToMatch, itemElement.Include, expandForXmlLocation(itemElement.IncludeLocation));
+
+            if (includeResult == null)
+            {
+                return null;
+            }
+
+            var excludeResult = ComputeProvenanceResult(itemToMatch, itemElement.Exclude, expandForXmlLocation(itemElement.ExcludeLocation));
```

Comment:
Looks like both uses of `expandForXmlLocation` use the same arg. Can you bake that in and avoid the currying?